packages/gatsby-link/src/LinkContainer.js

```diff
@@ -0,0 +1,185 @@
+      ___loader.enqueue(parsePath(this.props.to).pathname)
+    }
+  }
+
+  componentDidMount() {
+    // Preserve non IO functionality if no support
+    if (!this.state.IOSupported) {
+      ___loader.enqueue(parsePath(this.props.to).pathname)
+    }
+  }
+
+  handleRef(ref) {
+    this.props.innerRef && this.props.innerRef(ref)
+
+    if (this.state.IOSupported && ref) {
+      // If IO supported and element reference found, setup Observer functionality
+      handleIntersection(ref, () => {
+        ___loader.enqueue(parsePath(this.props.to).pathname)
+      })
+    }
+  }
+
+  defaultGetProps = ({ isCurrent }) => {
+    if (isCurrent) {
+      return {
+        className: [this.props.className, this.props.activeClassName]
+          .filter(Boolean)
+          .join(` `),
+        style: { ...this.props.style, ...this.props.activeStyle },
+      }
+    }
+    return null
+  }
+
+  render() {
+    const {
+      to,
+      getProps = this.defaultGetProps,
+      onClick,
+      onMouseEnter,
+      location,
+      /* eslint-disable no-unused-vars */
+      activeClassName: $activeClassName,
+      activeStyle: $activeStyle,
+      innerRef: $innerRef,
+      state,
+      children,
+      /* eslint-enable no-unused-vars */
+      ...rest
+    } = this.props
+
+    const prefixedTo = withPrefix(to)
+
+    return children({
+      state,
+      getProps,
+      to: prefixedTo,
+      ref: this.handleRef,
+      onMouseEnter: e => {
+        // eslint-disable-line
+        onMouseEnter && onMouseEnter(e)
+        ___loader.hovering(parsePath(to).pathname)
+      },
+      onClick: e => {
+        // eslint-disable-line
+        onClick && onClick(e)
+
+        if (
+          e.button === 0 && // ignore right clicks
+          !this.props.target && // let browser handle "target=_blank"
+          !e.defaultPrevented && // onClick prevented default
+          !e.metaKey && // ignore clicks with modifier keys...
+          !e.altKey &&
+          !e.ctrlKey &&
+          !e.shiftKey
+        ) {
+          e.preventDefault()
+          // Is this link pointing to a hash on the same page? If so,
+          // just scroll there.
+          const { pathname, hash } = parsePath(prefixedTo)
+          if (pathname === location.pathname || !pathname) {
+            const element = hash
+              ? document.getElementById(hash.substr(1))
+              : null
+            if (element !== null) {
+              element.scrollIntoView()
+            } else {
+              // This is just a normal link to the current page so let's emulate default
+              // browser behavior by scrolling now to the top of the page.
+              window.scrollTo(0, 0)
+            }
+          }
+
+          // Make sure the necessary scripts and data are
+          // loaded before continuing.
+          navigate(prefixedTo, { state })
+        }
+
+        return true
+      },
+      ...rest,
+    })
+  }
+}
+
+GatsbyLinkContainer.propTypes = propTypes
```

Comment:
propTypes here have required `location` which is not not there at exported component level - it get's added by `<Location>` in code below